lib/ansible/galaxy/dependency_resolution/providers.py

```diff
@@ -424,3 +478,63 @@ def get_dependencies(self, candidate):
+        return self._get_preference(candidates)
+
+
+class CollectionDependencyProvider060(CollectionDependencyProviderBase):
+    def find_matches(self, identifier, requirements, incompatibilities):  # type: ignore[override]
+        # type: (str, t.Mapping[str, t.Iterator[Requirement]], t.Mapping[str, t.Iterator[Requirement]]) -> list[Candidate]
+        return [
+            match for match in self._find_matches(list(requirements[identifier]))
+            if not any(match.ver == incompat.ver for incompat in incompatibilities[identifier])
+        ]
+
+    def get_preference(self, resolution, candidates, information):  # type: ignore[override]
+        # type: (Candidate | None, list[Candidate], list[t.NamedTuple]) -> float | int
+        return self._get_preference(candidates)
+
+
+class CollectionDependencyProvider070(CollectionDependencyProviderBase):
+    def find_matches(self, identifier, requirements, incompatibilities):  # type: ignore[override]
+        # type: (str, t.Mapping[str, t.Iterator[Requirement]], t.Mapping[str, t.Iterator[Requirement]]) -> list[Candidate]
+        return [
+            match for match in self._find_matches(list(requirements[identifier]))
+            if not any(match.ver == incompat.ver for incompat in incompatibilities[identifier])
+        ]
+
+    def get_preference(self, identifier, resolutions, candidates, information):  # type: ignore[override]
+        # type: (str, t.Mapping[str, Candidate], t.Mapping[str, t.Iterator[Candidate]], t.Iterator[t.NamedTuple]) -> float | int
+        return self._get_preference(list(candidates[identifier]))
+
+
+class CollectionDependencyProvider080(CollectionDependencyProviderBase):
+    def find_matches(self, identifier, requirements, incompatibilities):  # type: ignore[override]
+        # type: (str, t.Mapping[str, t.Iterator[Requirement]], t.Mapping[str, t.Iterator[Requirement]]) -> list[Candidate]
+        return [
+            match for match in self._find_matches(list(requirements[identifier]))
+            if not any(match.ver == incompat.ver for incompat in incompatibilities[identifier])
+        ]
+
+    def get_preference(self, identifier, resolutions, candidates, information, backtrack_causes):  # type: ignore[override]
+        # type: (str, t.Mapping[str, Candidate], t.Mapping[str, t.Iterator[Candidate]], t.Iterator[t.NamedTuple], t.Sequence) -> float | int
+        return self._get_preference(list(candidates[identifier]))
+
+
+if RESOLVELIB_080:
+    CollectionDependencyProvider = CollectionDependencyProvider080  # type: ignore[assignment,no-redef,misc]
+elif RESOLVELIB_070:
+    CollectionDependencyProvider = CollectionDependencyProvider070  # type: ignore[assignment,no-redef,misc]
+elif RESOLVELIB_060:
+    CollectionDependencyProvider = CollectionDependencyProvider060  # type: ignore[assignment,no-redef,misc]
+else:
+    CollectionDependencyProvider = CollectionDependencyProvider050  # type: ignore[assignment,no-redef,misc]
```

Comment:
I wonder if something like this would make mypy happier 🤷 

```suggestion
def _get_provider():  # type () -> CollectionDependencyProviderBase
    if RESOLVELIB_080:
        return CollectionDependencyProvider080
    elif RESOLVELIB_070:
        return CollectionDependencyProvider070
    elif RESOLVELIB_060:
        return CollectionDependencyProvider060
    return  CollectionDependencyProvider050


CollectionDependencyProvider = _get_provider()
```

Reply:
Yes, it seems to like that much better :smile: